### PR TITLE
POST/ processes endpoint

### PR DESCRIPTION
**Pull request by geomatician**
POST/processes endpoint for SWOOP

### CODE REVIEW

File: src/swoop/api/routers/processes.py [hunk 1]
```diff
@@ -138,17 +149,64 @@ async def get_process_definition(process_id: str = Path(..., alias="processID"))
...
+        )
+        await conn.fetchrow(q, *p)
+
+        q, p = render(
+            f"""
+                INSERT INTO swoop.action(action_uuid, action_type,
+                    action_name, handler_name, parent_uuid, payload_uuid)
+                VALUES ('{action_uuid}','workflow','{workflow.name}',
+                '{workflow.handler}','{uuid.uuid4()}','{pl_uuid}');
+            """,
```
Comment (jkeifer) [inline src/swoop/api/routers/processes.py:208]: If you find yourself using string templating for sql queries, _stop immediately_.

Please look at the buildpg documentation and other API queries for examples of how to write queries that are not vulnerable to sql injection.

File: src/swoop/api/workflows.py [hunk 2]
```diff
@@ -1,7 +1,11 @@
...
+from swoop.api.models.workflows import Workflows
+from swoop.api.routers.processes import create_workflows_dict
+
 
 def init_workflows_config(app: FastAPI) -> None:
     """Initialize Workflow Config."""
-    app.state.workflows = yaml.safe_load(open(app.state.settings.workflow_config_file))
+
+    wf = Workflows.from_yaml(app.state.settings.workflow_config_file)
+    app.state.workflows = create_workflows_dict(wf)
```
Comment (jkeifer) [inline src/swoop/api/workflows.py:11]: Please see my PR into your branch for a better way to do this: https://github.com/Element84/swoop/pull/23.

File: src/swoop/api/routers/processes.py [hunk 3]
```diff
@@ -28,22 +26,20 @@ class Params(BaseModel):
...
-        cacheKeyHashIncludes=workflowConfig.get("cache_key_hash_includes"),
-        cacheKeyHashExcludes=workflowConfig.get("cache_key_hash_excludes"),
+        id=workflowConfig.name,
+        title=workflowConfig.name,
+        name=workflowConfig.name,
+        processID=workflowConfig.name,
+        version=workflowConfig.version,
+        description=workflowConfig.description,
+        handler=workflowConfig.handler,
+        # argoTemplate=workflowConfig.workflow_template,
```
Comment (jkeifer) [inline src/swoop/api/routers/processes.py:40]: The current submodel-specific fields `argo_template` and `sfn_arn` I don't think need to be exposed to end users.

File: src/swoop/api/routers/processes.py [hunk 4]
```diff
@@ -72,16 +84,19 @@ async def list_processes(
...
     if queryparams and len(workflows) > 0:
         workflows = list(
             filter(
-                lambda x: queryparams.items() <= x.items(),
-                request.app.state.workflows["workflows"],
+                lambda x: queryparams.items() <= vars(x).items(),
+                workflows,
             )
         )
+
```
Comment (jkeifer) [inline src/swoop/api/routers/processes.py:99]: I'm not sure if I see the reason to filter workflows like this. Maybe we can discuss in parking lot whether this is valuable and warrants the additional complexity.

File: src/swoop/api/routers/processes.py [hunk 5]
```diff
@@ -112,20 +127,16 @@ async def get_process_description(
     workflows = (
         request.app.state.workflows.get("workflows")
         if request.app.state.workflows.get("workflows")
-        else []
+        else {}
```
Comment (jkeifer) [inline src/swoop/api/routers/processes.py:130]: I don't think this is right. I think it is just `request.app.state.workflows`, which will always be a dict.

**Review by jkeifer — CHANGES_REQUESTED** [review]
Please rebase your changes off the main branch now that the io client changes are merged.

Please review and integrate my approach to the workflow model instantiation, then review and fix all tests.

Please rework the sql queries as I mentioned.

Now that you refactored `models.py` into a package, please remove the `models.py module.

File: tests/api/test_processes.py [hunk 6]
```diff
@@ -179,3 +227,35 @@ async def test_get_process_by_process_id(test_client, mirror_workflow_process):
...
     response = test_client.get("/processes/mirror-test")
     assert response.status_code == 404
+
+
+@pytest.mark.asyncio
+async def test_post_valid_name_valid_payload(test_client, process_payload_valid):
+    response = test_client.post(
+        "/processes/mirror/execution", data=json.dumps(process_payload_valid)
+    )
+    assert response.status_code == 200
```
Comment (jkeifer) [inline tests/api/test_processes.py:237]: Per the spec, we should be returning a 201 for an async response (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201).

Comment (geomatician) [reply tests/api/test_processes.py]: Fixed

File: src/swoop/api/models/workflows.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    @classmethod
+    def from_yaml(cls, path: Path) -> Workflows:
+        try:
+            workflows = yaml.safe_load(path.read_text())["workflows"]
+            for name, workflow in workflows.items():
+                workflow["name"] = name
+            return cls(cls._type.parse_obj(workflows).__root__)
+        except Exception as e:
+            raise WorkflowConfigError("Could not load workflow configuration") from e
```
Comment (jkeifer) [inline src/swoop/api/models/workflows.py:101]: We probably ought to have at least one test of a bad workflow config or something to trigger this exception.

File: src/swoop/api/routers/processes.py [hunk 8]
```diff
@@ -109,46 +95,83 @@ async def get_process_description(
...
+        raise HTTPException(status_code=404, detail="Process not found")
+
+    # Generate a UUID
+    action_uuid = uuid.uuid4()
+
+    # Write to object storage
+
+    request.app.state.io.put_object(
+        object_name=f"executions/{action_uuid}/input.json",
+        object_content=json.dumps(body.dict()["inputs"]),
```
Comment (jkeifer) [inline src/swoop/api/routers/processes.py:145]: I think this should actually be `body.dict()["inputs"]["payload"]`, right?

Perhaps you want to pull the payload out up above, so you have it to reference in the couple places you need the payload?

File: src/swoop/api/routers/processes.py [hunk 9]
```diff
@@ -109,46 +95,83 @@ async def get_process_description(
...
-) -> InlineResponse200 | StatusInfo | APIException:
+async def execute_process(
+    process_id: str, request: Request, body: Execute
+) -> StatusInfo | APIException:
     """
     execute a process.
     """
-    pass
+
+    wf_name = body.dict().get("inputs").get("payload").get("process")[0].get("workflow")
```
Comment (jkeifer) [inline src/swoop/api/routers/processes.py:125]: This line assumes `process` in the payload is an array, but the model supports `process` being an array or a map.

After further discussion with @matthewhanson, we reached the conclusion that we will no longer support the map, and `process` should always be an array. TBD is whether this should now be renamed to `processes` or something, but for now I think just change the model so `process` is always an array and this is fine.

**Review by jkeifer — CHANGES_REQUESTED** [review]

File: src/swoop/api/routers/processes.py (hunk 8 above)
Comment (geomatician) [reply src/swoop/api/routers/processes.py]: Fixed

File: src/swoop/api/routers/processes.py (hunk 9 above)
Comment (geomatician) [reply src/swoop/api/routers/processes.py]: Fixed

**Review by jkeifer — CHANGES_REQUESTED** [review]

**Review by jkeifer — APPROVED** [review]

**Pull request merged**